keps/sig-network/20191220-apiserver-url-envvar.md

```diff
@@ -0,0 +1,294 @@
+naively concatenated into a URL when `KUBERNETES_SERVICE_HOST` is an
+IPv6 literal address.  As introduced by [RFC 2732], IPv6 literal
+addresses must be escaped with square brackets when forming a URL to
+avoid ambiguity with URL port numbers - for example
+`http://[1080::8:800:200C:417A]:6443`.  This is unlike IPv4 literals
+or hostnames (even when hostnames resolve to IPv6 addresses), which
+cannot themselves contain colon characters.
+
+[RFC 2732]: https://tools.ietf.org/html/rfc2732
+
+Concatenating the `KUBERNETES_SERVICE_HOST` and `..._PORT`
+_incorrectly_ is a common error.  See for example the [java], [ruby],
+[haskell], and [javascript] clients.  This is particularly difficult
+in shell scripts, where there is no helper library available.
+
+[java]: kubernetes-client/java#839
+[ruby]: kubernetes-client/ruby#51
+[haskell]: kubernetes-client/haskell#68
+[javascript]: kubernetes-client/javascript#380
+
+Secondarily, the general `FOO_SERVICE_HOST` and `FOO_SERVICE_PORT`
+environment variables are a docker-compose-era legacy that works pooly
+in Kubernetes, and we would like to [remove support for them one
+day][servicevarskep].  Replacing the in-cluster Kubernetes API
+discovery logic removes the last "need" for this family of environment
+variables.
+
+[servicevarskep]: /keps/sig-apps/0028-20180925-optional-service-environment-variables.md
+
+### Goals
+
+- Provide a pointer to the in-cluster API server endpoint that "just
+  works" for IPv4, IPv6 and hostname forms.
+
+### Non-Goals
+
+- Actually remove `FOO_SERVICE_HOST`/`FOO_SERVICE_PORT` variables.
+  This is covered by [other KEPs][servicevarskep].
+
+## Proposal
+
+1. Unconditionally inject a new environment variable into pods called
+   `KUBERNETES_SERVICE_URL`.  The value would be the correctly-escaped
+   equivalent of
+   `https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT`. Note
+   that this would continue to use a (single) IPv4 or v6 literal
+   address in most cases.
+
+   An open question is whether the URL could ever include a path
+   component.  This KEP makes no statement about that, but notes the
+   choice of URL format leaves this and alternate schemes as possible
+   extension points in future (unlike just "host:port" format).
+
+2. Announce the deprecation of the existing `KUBERNETES_SERVICE_HOST`
+   and `KUBERNETES_SERVICE_PORT`.  Given the widespread client impact
+   this KEP does not propose a timeline on actual removal, just that
+   in-cluster clients SHOULD use the new `KUBERNETES_SERVICE_URL`
+   mechanism.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a client developer, I would like to just use a discovery mechanism
+without having to think about IPv6 syntax details.
+
+#### Story 2
+
+As a cluster admin, I would like to enable an apiserver IPv6 endpoint
+without encountering widespread URL bugs.
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation?
```

Comment:
What if a pod was previously defining an environment variable named `KUBERNETES_SERVICE_URL` but with different semantics? (eg, it points to something other than the apiserver)

Reply:
We can only set it if we don't find any existing `env` of the same name.

Reply:
That doesn't fully solve the problem though. If we think there actually might be pods defining a `KUBERNETES_SERVICE_URL` variable (or whatever we call it) and using it for something other than pointing to the apiserver, and we're not going to override its value in that case, then that means that generic code running in a pod can't assume that `KUBERNETES_SERVICE_URL` necessarily contains a URL pointing to the apiserver.

The easiest thing to do would be to assume/pretend this isn't actually going to happen (possibly after picking an even-less-likely-to-conflict name for the variable), and if it later turns out that someone was using that name and things break, then apologize and tell them they have to rename their variable.

Alternatively, we could say that the new variable really only exists for use directly from the `command` of a `PodSpec`, where the person using the variable can see that the pod isn't overriding it to mean something else, and that generic code like `rest.InClusterConfig` that may end up running in arbitrary pods needs to keep using `KUBERNETES_SERVICE_HOST` / `KUBERNETES_SERVICE_PORT`. This implies we continue supporting the old variables forever, and `KUBERNETES_SERVICE_URL` can never be anything besides just the correctly-concatenated value of the two old variables (eg, no `ExternalName`).

Or, if we take a bit longer to roll things out, we could declare that in the future, all environment variables starting with `KUBERNETES_` will be reserved for kubernetes-internal usage and can't be used by pods for unrelated purposes. To distinguish "I'm coincidentally using a conflicting variable name" from "I'm intentionally overriding a standard variable", we'd add an `override` field to `corev1.EnvVar`, and make kubelet throw warnings for the next 2 releases any time it launches a pod that sets a variable with a name starting with `KUBERNETES_` without setting `override: true` (and warn anyway if they're "overriding" a variable that kubelet doesn't know about). Then after 2 releases, make it an error rather than a warning to omit `override` (but still just a warning to override an unknown variable), and introduce the new apiserver URL variable.

Reply:
Dan makes a good point.  There's literally no name we can assume is not in use, and WAY back, we opted not to reserve names with a prefix.  I am still not comfortable reserving a prefix in a space that is really the end-users' to own.

I think "set-unless-already-set" is the best we can do, maybe with an event or something to indicate a warning.  That does mean that the old variables are there for a very long time, and that is also part of why ExternalName is a hard transition to make.  We can fix all the places internally that do dumb catenation of host and port to be v6 safe.  We can maybe even use the URL if it exists AND "tastes right" (for some definition).